Feature #2781 PairStat wrapper

georgemccabe opened the pull request:
## Pull Request Testing ##

- [X] Describe testing already performed for these changes:</br>

- [X] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

Review code changes, ensure all tests pass (exceptions described below)

- [X] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes]**

- [X] Do these changes include sufficient testing updates? **[Yes]**

- [X] Will this PR result in changes to the test suite? **[Yes]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>

Most of differences resulting from this pull request are expected. This PR includes refactoring to use the same function to write file list files for the CompareGriddedWrappers (GridStat, PointStat, EnsembleStat, MODE, MTD, and PairStat). Also, the new PairStat basic use case created new output data.

There are differences in one s2s use case that should be investigated.

[met_tool_wrapper:0-29,59-64](https://github.com/dtcenter/METplus/actions/runs/12955526919/job/36141862860#step:9:2259)
Rename file list files, e.g. 
stage/file_lists/20091231120000_24_ensemble_stat.txt
to
stage/file_lists/ensemble_stat_files_FCST_init_20091231120000_valid_20100101120000_lead_86400.txt

[met_tool_wrapper:66](https://github.com/dtcenter/METplus/actions/runs/12955526919/job/36141865009#step:9:14)
Truth output does not yet exist in DockerHub. We could improve the logging so it is more clear that this is why it fails

[air_quality_and_comp:0](https://github.com/dtcenter/METplus/actions/runs/12955526919/job/36141865526#step:9:124)
Rename file list file
stage/file_lists/20160815000000_12_ensemble_stat.txt
to
stage/file_lists/ensemble_stat_files_FCST_init_20160815000000_valid_20160815120000_lead_43200.txt

[preciptation:3-7](https://github.com/dtcenter/METplus/actions/runs/12955526919/job/36141874355#step:9:583)
Rename file list file
stage/file_lists/20200615170000_1_ensemble_stat.txt
to
stage/file_lists/ensemble_stat_files_FCST_init_20200615170000_valid_20200615180000_lead_3600.txt

[s2s_stratosphere:0](https://github.com/dtcenter/METplus/actions/runs/12955526919/job/36141878136#step:9:733)
Rename file lists files, e.g.
series_analysis_files_fcst_init_ALL_valid_ALL_lead_ALL.txt
and
series_analysis_files_obs_init_ALL_valid_ALL_lead_ALL.txt
to
series_analysis_files_fcst_init_ALL_valid_20180201000000_lead_ALL.txt
and
series_analysis_files_obs_init_ALL_valid_20180201000000_lead_ALL.txt

[short_range:0](https://github.com/dtcenter/METplus/actions/runs/12955526919/job/36141878408#step:9:422)
Rename file list files, e.g.
stage/file_lists/20180709120000_2_ensemble_stat.txt
to
stage/file_lists/ensemble_stat_files_FCST_init_20180709120000_valid_20180709140000_lead_7200.txt

[short_range:13](https://github.com/dtcenter/METplus/actions/runs/12955526919/job/36141879980#step:9:193)
Rename file list files
stage/file_lists/20210201000000_21_mode_fcst.txt
and
stage/file_lists/20210201000000_21_mode_obs.txt
to
stage/file_lists/mode_files_FCST_init_20210201000000_valid_20210201210000_lead_75600.txt
and
stage/file_lists/mode_files_OBS_init_20210201000000_valid_20210201210000_lead_75600.txt

- [X] Do these changes introduce new SonarQube findings? **[No]**</br>
If **yes**, please describe:

- [X] Please complete this pull request review by **1/27/2025**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [X] Add any new Python packages to the [METplus Components Python Requirements](https://metplus.readthedocs.io/en/develop/Users_Guide/appendixA.html#metplus-components-python-packages) table.
- [X] For any new datasets, an entry to the [METplus Verification Datasets Guide](https://metplus.readthedocs.io/en/latest/Verification_Datasets/index.html).
- [X] Review the source issue metadata (required labels, projects, and milestone).
- [X] Complete the PR definition above.
- [X] Ensure the PR title matches the feature or bugfix branch name.
- [X] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issue
Select: **Milestone** as the version that will include these changes
Select: **Coordinated METplus-X.Y Support** project for bugfix releases or **METplus-Wrappers-X.Y.Z Development** project for official releases
- [X] After submitting the PR, select the :gear: icon in the **Development** section of the right hand sidebar. Search for the issue that this PR will close and select it, if it is not already selected.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

georgemccabe committed:
update release guide for METplus to create data location for next release on mohawk when creating rc1 release instead of after official release

georgemccabe committed:
cleanup

georgemccabe committed:
per #2781, start first implementation of PairStat wrapper

georgemccabe committed:
merge develop and resolve conflicts

georgemccabe committed:
appease SonarQube by increasing code coverage for component versions script

georgemccabe committed:
add new wrapper to lookup dictionary

georgemccabe committed:
per #2781 add wrapped MET config file

georgemccabe committed:
use runtime freq logic to find input files

georgemccabe committed:
start unit tests (broken)

georgemccabe committed:
Merge branch 'develop' into feature_2781_pair_stat

georgemccabe committed:
reorder assert to align expected and actual values with PyCharm notation

georgemccabe committed:
update unit tests

georgemccabe committed:
rename field info variable 'field' to 'pairs' to match Pair-Stat naming convention. Add command line arguments to pair_stat command

georgemccabe committed:
reorder assert to align actual and expected in PyCharm

georgemccabe committed:
refactor function to get start/end times to prevent error when using INIT/VALID_LIST and use time_generator to be more consistent and reduce duplicate code

georgemccabe committed:
change many wrappers to be consistent with finding input files. Errors are now thrown when any input file is not found, checking other input types even when another input was not found. This increased errors reported in unit tests, so updated tests to reflect this.

georgemccabe committed:
remove line that is not needed

georgemccabe committed:
fix bug introduced with recent changes and run use cases that failed to test that the fix is correct

georgemccabe committed:
fix multi-variate MODE run to pass all fields to the call instead of just the first one

georgemccabe committed:
update SeriesAnalysis test to use logic that is actually used when running

georgemccabe committed:
turn off use cases that now succeed and turn on use case to test fix to multi-variate mode

georgemccabe committed:
turn off use case after confirming that it now runs successfully

georgemccabe committed:
refactor logic to satisfy SonarQube to reduce cognitive complexity and remove unused imports

georgemccabe committed:
remove functions that are not used

georgemccabe committed:
handle -pairs flag like other command line args for input files. Add documentation to new functions

georgemccabe committed:
updates to new wrapper, add documentation, basic use case stubs. create function for duplicate code to handle land_mask and topo_mask in PointStat and PairStat wrappers

georgemccabe committed:
remove unused import

georgemccabe committed:
remove config variable that is no longer used in MET

georgemccabe committed:
remove match_month support from climo_mean/stdev

georgemccabe committed:
remove obs_valid_beg/end from command line args because it is not supported by pair_stat

georgemccabe committed:
remove test config files that are not being used

georgemccabe committed:
configure basic use case to run command that is being used to test/develop the app

georgemccabe committed:
error if invalid tool name is provided

georgemccabe committed:
add missing tests

georgemccabe committed:
remove support for config variables that did not make it into the final implementation of the MET pair_stat tool

georgemccabe committed:
fix formatting on a few config values

georgemccabe committed:
correct input file info

georgemccabe committed:
turn on use case to test

georgemccabe committed:
merged develop and resolved conflicts

georgemccabe committed:
fix command line argument for output directory that changed since the last test

georgemccabe committed:
refactor and clean up to reduce SonarQube issues

JohnHalleyGotway commented:
@georgemccabe we have differences flagged for many use cases in this [testing workflow run](https://github.com/dtcenter/METplus/actions/runs/12955526919). Can you please take a look and advise as to whether or not these diffs are expected for the changes proposed by this pull request?

JohnHalleyGotway commented on docs/Users_Guide/glossary.rst:
```diff
@@ -13319,6 +13269,772 @@ METplus Configuration Glossary
+
+     | *Used by:* PairStat
+
+   PAIR_STAT_OUTPUT_PREFIX
```
Please remove `PAIR_STAT_OUTPUT_PREFIX` which is not used anymore since `-outdir directory` was replaced with `-out file_name_base` on the command line.

JohnHalleyGotway commented on docs/Users_Guide/glossary.rst:
```diff
@@ -13319,6 +13269,772 @@ METplus Configuration Glossary
+
+     | *Used by:*  PairStat
+
+   PAIR_STAT_OUTPUT_TEMPLATE
```
Just want to confirm that this is used to set the `-out file_name_base` command line option and not the `-oudir directory` command line option that is no longer used.

JohnHalleyGotway commented on docs/Users_Guide/glossary.rst:
```diff
@@ -13319,6 +13269,772 @@ METplus Configuration Glossary
+     | *Used by:*  PairStat
+
+   PAIR_STAT_OUTPUT_TEMPLATE
+     Sets the subdirectories below :term:`PAIR_STAT_OUTPUT_DIR` using a template to allow run time information.
```
```suggestion
     Sets the subdirectories and output file name base below :term:`PAIR_STAT_OUTPUT_DIR` using a template to allow run time information.
```

JohnHalleyGotway commented on docs/Users_Guide/glossary.rst:
```diff
@@ -13319,6 +13269,772 @@ METplus Configuration Glossary
 
      | *Used by:* WaveletStat
 
```
Please edit the entry for `MODEL` to add `PairStat` to the list of wrappers that use that entry.

JohnHalleyGotway commented on docs/Users_Guide/glossary.rst:
```diff
@@ -13319,6 +13269,772 @@ METplus Configuration Glossary
+
+     | *Used by:* PairStat
+
+   PAIR_STAT_OBS_WINDOW_BEG
```
Side note, the obs window is being parsed from the Pair-Stat configuration file but not actually USED in the initial version to filter pairs by time yet. Good to have in the wrapper for now, but there's potential for this to changes as the tool matures.

JohnHalleyGotway commented on docs/Users_Guide/wrappers.rst:
```diff
@@ -5923,6 +5905,790 @@ ${METPLUS_MET_CONFIG_OVERRIDES}
+| :term:`PAIR_STAT_PAIRS_INPUT_DIR`
+| :term:`PAIR_STAT_PAIRS_INPUT_DATATYPE`
+| :term:`PAIR_STAT_OUTPUT_TEMPLATE`
+| :term:`PAIR_STAT_OUTPUT_DIR`
```
Since `-outdir directory` is replaced with `-out filename base`, `PAIR_STAT_OUTPUT_DIR` should probably be `PAIR_STAT_OUTPUT_BASE` instead? For example `-out path/to/abc` produces output files named `path/to/abc.stat`, `path/to/abc_mpr.txt`, and so on. So this is same output file logic used for TC-Pairs and TC-Gen tools.

JohnHalleyGotway commented on internal/tests/pytests/wrappers/pair_stat/test_pair_stat_wrapper.py:
```diff
@@ -0,0 +1,756 @@
+    config.set('config', 'PAIR_STAT_PAIRS_INPUT_DIR', pairs_dir)
+    config.set('config', 'PAIR_STAT_PAIRS_INPUT_TEMPLATE',
+               '{init?fmt=%Y%m%d%H}/fcst_file_F{lead?fmt=%3H}')
+    config.set('config', 'PAIR_STAT_OUTPUT_DIR',
```
Possible impact here related to Pair-Stat's use of `-out` on the command line rather than `-outdir`.

georgemccabe committed:
update use case to prevent redundant runs of SeriesAnalysis(run_two) by only setting custom loop list for the first instance of SeriesAnalysis

georgemccabe committed:
Fix bug that prevents correct field information from being set when multiple fields are read from the same file. Removed unused function

georgemccabe committed:
per feedback in PR #2872, remove support for setting output_prefix

georgemccabe committed:
add PairStat to list of wrappers that use the MODEL config variable

georgemccabe committed:
per feedback in PR #2872, modify wrapper behavior to change output path from a directory to a filename base that will be used to write output files with various extensions, e.g. _mpr.txt or .stat

georgemccabe committed:
remove output_prefix from tables

georgemccabe commented:
@JohnHalleyGotway, I made updates to the PairStat wrapper based on your suggestions. I was testing with an older version of the MET tool. The wrapper now handles the output file name base properly. Other wrappers that write files this way, like TCGen, use the *_OUTPUT_DIR and *_OUTPUT_TEMPLATE naming convention. I removed support for setting `output_prefix`.

I also fixed a bug in SeriesAnalysis that was preventing the correct field information to be set in regards to time, e.g. when data is in a single input file and the field information dictates what to read from the file.

The differences in output are expected and are described in the PR body.

JohnHalleyGotway reviewed (approved):
I approve of these changes. Thanks for all your work on this @georgemccabe!

- I downloaded the diff artifacts and confirmed that they're limited to ASCII file lists, which is fine and expected.

- I downloaded the new met_tool_wrapper:66 output and confirmed that it does contain Pair-Stat output.

- I stepped through the changes in 51 files and note the following:

  - met_tool_wrapper:66 is a new use case group for which output does not yet exist. This is for Pair-Stat tests:```66::PairStat:: met_tool_wrapper/PairStat/PairStat.conf```

- Functionality changes:
  - `MATCH_MONTH` is fully deprecated throughout, in the code and the docs.
- Doc changes:
  - `glossary.rst` adds `PAIR_STAT` entries. Reviewed contents for accuracy and made a couple suggestions.

- Config files removed (not entirely clear why but all the test run so that's fine)
    - GridStatConfig_prob_precip
    - STATAnalysisConfig
    - ascii2nc_mask_grid.conf, ascii2nc_mask_poly.conf, ascii2nc_mask_sid.conf, ascii2nc_multi_file.conf, ascii2nc_single_file.conf, grid_gefs_prob.conf 

- Updates to pytests related to removing MATCH_MONTH.
- Consolidation and standardization of time looping logic.
